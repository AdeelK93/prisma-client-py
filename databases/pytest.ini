[pytest]
# asyncio fixtures and tests must be explicitly marked
asyncio_mode = auto
