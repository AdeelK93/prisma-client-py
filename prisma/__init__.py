# -*- coding: utf-8 -*-

__title__ = 'prisma'
__author__ = 'RobertCraigie'
__license__ = 'APACHE'
__copyright__ = 'Copyright 2020 RobertCraigie'
__version__ = '0.0.1'

from .plugins import *
from . import binaries, jsonrpc, engine, utils


try:
    from .client import *
except ModuleNotFoundError:
    # code has not been generated yet
    # TODO: this could swallow unexpected errors
    pass


utils.setup_logging()
