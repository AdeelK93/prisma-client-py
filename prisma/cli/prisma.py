import os
import sys
import logging
import subprocess
from typing import List, Union, Optional, IO

from .. import generator, jsonrpc, binaries


log = logging.getLogger(__name__)


def run(
    args: List[str],
    check: bool = False,
    pipe: bool = False,
) -> int:
    directory = binaries.ensure_cached()
    path = directory.joinpath(binaries.PRISMA_CLI_NAME)
    if not path.exists():
        raise RuntimeError(
            f'The Prisma CLI is not downloaded, expected {path} to exist.'
        )

    log.debug('Using Prisma CLI at %s', path)
    log.debug('Running prisma command with args: %s', args)

    env = {'PRISMA_HIDE_UPDATE_MESSAGE': 'true', **os.environ}

    # ensure the client uses our engine binaries
    for engine in binaries.ENGINES:
        env[engine.env] = str(engine.path.absolute())

    encoding = None  # type: Optional[str]
    stdout = sys.stdout  # type: Union[int, IO[str]]
    stderr = sys.stderr  # type: Union[int, IO[str]]

    if pipe:
        stdout = subprocess.PIPE
        stderr = subprocess.PIPE
        encoding = 'utf-8'

    process = subprocess.run(
        [str(path.absolute()), *args],
        env=env,
        check=check,
        stdout=stdout,
        stderr=stderr,
        encoding=encoding,
    )

    if pipe:
        print(process.stdout)
        print(process.stderr, file=sys.stderr)

    return process.returncode


def invoke() -> None:
    while True:
        line = jsonrpc.readline()
        if line is None:
            log.debug('Prisma invocation ending')
            break

        request = jsonrpc.parse(line)
        log.debug(
            'Received request method: %s',
            request.method,
        )

        # TODO: this can hang the prisma process if an error occurs
        response = None
        if request.method == 'getManifest':
            response = jsonrpc.Response(
                id=request.id,
                result=dict(
                    manifest=jsonrpc.Manifest(
                        defaultOutput=str(generator.BASE_PACKAGE_DIR.absolute()),
                        prettyName='Prisma Client Python',
                    )
                ),
            )
        elif request.method == 'generate':
            if request.params is None:
                raise RuntimeError('Prisma JSONRPC did not send data to generate.')

            generator.run(request.params)
            response = jsonrpc.Response(id=request.id, result=None)
        else:
            raise RuntimeError(f'JSON RPC received unexpected method: {request.method}')

        if response is not None:
            jsonrpc.reply(response)
