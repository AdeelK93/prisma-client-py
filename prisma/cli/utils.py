import sys
import logging
from pathlib import Path
from subprocess import CompletedProcess
from typing import Any  # pylint: disable=unused-import
from typing import Optional, List, Union, NoReturn, overload

import click

from ..utils import module_exists
from .._types import Literal


log = logging.getLogger(__name__)


class PrismaCLI(click.MultiCommand):

    base_package = 'prisma.cli.commands'
    folder = Path(__file__).parent / 'commands'

    def list_commands(self, ctx: click.Context) -> List[str]:
        commands = []

        for path in self.folder.iterdir():
            name = path.name
            if name.startswith('_'):
                continue

            if name.endswith('.py'):
                commands.append(path.stem)
            elif is_module(path):
                commands.append(name)

        commands.sort()
        return commands

    def get_command(self, ctx: click.Context, cmd_name: str) -> Optional[click.Command]:
        name = f'{self.base_package}.{cmd_name}'
        if not module_exists(name):
            # command not found
            return None

        mod = __import__(name, None, None, ['cli'])

        assert hasattr(
            mod, 'cli'
        ), f'Expected command module {name} to contain a "cli" attribute'
        assert isinstance(mod.cli, click.Command), (
            f'Expected command module attribute {name}.cli to be a {click.Command} '
            f'instance but got {type(mod.cli)} instead'
        )

        return mod.cli


def is_module(path: Path) -> bool:
    return path.is_dir() and path.joinpath('__init__.py').exists()


@overload
def error(message: str) -> NoReturn:
    ...


@overload
def error(message: str, exit_: Literal[True]) -> NoReturn:
    ...


@overload
def error(message: str, exit_: Literal[False]) -> None:
    ...


def error(message: str, exit_: bool = True) -> Union[None, NoReturn]:
    click.echo(click.style(message, fg='bright_red', bold=True), err=True)
    if exit_:
        sys.exit(1)
    else:
        return None
