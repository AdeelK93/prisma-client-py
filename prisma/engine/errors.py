from ..errors import PrismaError


__all__ = (
    'EngineError',
    'BinaryNotFoundError',
    'MismatchedVersionsError',
    'EngineConnectionError',
    'EngineRequestError',
    'AlreadyConnectedError',
)


class EngineError(PrismaError):
    pass


class BinaryNotFoundError(EngineError):
    pass


class AlreadyConnectedError(EngineError):
    pass


class MismatchedVersionsError(EngineError):
    def __init__(self, message=None, *, expected=None, got=None):
        super().__init__(
            message or f'Expected query engine version `{expected}` but got `{got}`'
        )
        self.expected = expected
        self.got = got


class EngineConnectionError(EngineError):
    pass


class EngineRequestError(EngineError):
    def __init__(self, response, body):
        self.response = response

        # TODO: better error message
        super().__init__(f'{response.status}: {body}')
