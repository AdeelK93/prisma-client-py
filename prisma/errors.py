from typing import Any


__all__ = (
    'PrismaError',
    'DataError',
    'UniqueViolationError',
)


class PrismaError(Exception):
    pass


class DataError(PrismaError):
    def __init__(self, data: Any):
        self.data = data

        user_facing_error = data.get('user_facing_error', {})
        self.code = user_facing_error.get('error_code')
        self.meta = user_facing_error.get('meta')

        message = user_facing_error.get('message')
        super().__init__(message or 'An error occurred while processing data.')


class UniqueViolationError(DataError):
    pass


class MissingRequiredValueError(DataError):
    pass
