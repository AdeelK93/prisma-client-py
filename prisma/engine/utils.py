import os
import time
import socket
import logging
import subprocess
from pathlib import Path
from typing import NoReturn, Any

import aiohttp

from . import errors
from .. import errors as prisma_errors

from ..utils import time_since
from ..binaries import GLOBAL_TEMP_DIR, ENGINE_VERSION, platform


log = logging.getLogger(__name__)


def ensure() -> Path:
    start_time = time.monotonic()
    file = None
    force_version = True

    # TODO: add support for exact binary names
    # for example "linux-openssl-1.1.x" instead of "linux"
    binary_name = platform.check_for_extension(platform.binary_platform())

    name = f'prisma-query-engine-{binary_name}'
    local_path = Path.cwd().joinpath(name)
    global_path = GLOBAL_TEMP_DIR.joinpath(name)

    log.debug('Expecting local query engine %s', local_path)
    log.debug('Expecting global query engine %s', global_path)

    binary = os.environ.get('PRISMA_QUERY_ENGINE_BINARY')
    if binary:
        log.debug('PRISMA_QUERY_ENGINE_BINARY is defined, using %s', binary)

        if not Path(binary).exists():
            raise errors.BinaryNotFoundError(
                'PRISMA_QUERY_ENGINE_BINARY was provided, '
                f'but no query engine was found at {binary}'
            )

        file = Path(binary)
        force_version = False
    elif local_path.exists():
        file = local_path
        log.debug('Query engine found in the working directory')
    elif global_path.exists():
        file = global_path
        log.debug('Query engine found in the global path')

    if not file:
        raise errors.BinaryNotFoundError()

    start_version = time.monotonic()
    process = subprocess.run(
        [file.absolute(), '--version'], stdout=subprocess.PIPE, check=True
    )
    log.debug('Version check took %s', time_since(start_version))

    version = str(process.stdout, 'utf-8').replace("query-engine", "").strip()
    log.debug('Using query enginve version %s', version)

    if force_version and version != ENGINE_VERSION:
        raise errors.MismatchedVersionsError(expected=ENGINE_VERSION, got=version)

    log.debug('Using query engine at %s', file)
    log.debug('Ensuring query engine took: %s', time_since(start_time))

    return file


def get_open_port() -> int:
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.bind(('', 0))
    port = sock.getsockname()[1]
    sock.close()
    return int(port)


def handle_response_errors(resp: aiohttp.ClientResponse, data: Any) -> NoReturn:
    for error in data:
        try:
            code = error.get('user_facing_error', {}).get('error_code')
            if code is None:
                continue

            if code == 'P2002':
                raise prisma_errors.UniqueViolationError(error)

            if code == 'P2010':
                raise prisma_errors.RawQueryError(error)

            if code == 'P2012':
                raise prisma_errors.MissingRequiredValueError(error)

            if code == 'P2021':
                raise prisma_errors.TableNotFoundError(error)
        except (KeyError, TypeError):
            continue

    try:
        raise prisma_errors.DataError(data[0])
    except (IndexError, TypeError):
        pass

    raise errors.EngineRequestError(
        resp, f'Could not process erroneous response: {data}'
    )
