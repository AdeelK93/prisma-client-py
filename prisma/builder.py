import json
from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field

from .engine import QueryEngine


class QueryBuilder(BaseModel):
    # prisma method
    method: str

    # GraphQL operation
    operation: str

    model: Optional[str]
    engine: QueryEngine
    arguments: Dict[str, Any]
    fields_: Optional[List[str]] = Field(alias='fields')

    class Config:
        arbitrary_types_allowed = True

    def build(self) -> str:
        data = {
            'variables': {},
            'operation_name': self.operation,
            'query': self.build_query(),
        }
        return json.dumps(data)

    def build_query(self) -> str:
        """Build the GraphQL query string for this operation.

        e.g the following without newlines

        mutation {
          result: createOnePost(
            data: {
              title: "Hi from Prisma!"
              published: true
              desc: "Prisma is a database toolkit that makes databases easy."
            }
          ) { id published createdAt updatedAt title desc }
        }
        """
        # fmt: off
        return (
            f'{self.operation} {{'
              f'result: {self.method}{self.model if self.model else ""}('
                f'{self.build_args()}'
              f'){self.build_fields()}'
            '}'
        )
        # fmt: on

    def build_args(self) -> str:
        """Returns a string representing the arguments to the the GraphQL query.

        e.g. 'where: { id: "hsdajkd2" } take: 1'
        """
        strings = []
        for arg, value in self.arguments.items():
            if isinstance(value, dict):
                parsed = self.build_data(value)
            elif isinstance(value, (list, tuple)):
                # prisma expects a json string value like "[\"John\",\"123\"]"
                # we encode twice to ensure that only the inner quotes are escaped
                # yes this is a terrible solution
                # TODO: better solution
                parsed = json.dumps(json.dumps(value))
            else:
                parsed = json.dumps(value)

            strings.append(f'{arg}: {parsed}')

        return ' '.join(strings)

    def build_fields(self) -> str:
        if self.fields_ is None:
            return ''

        return f'{{ {" ".join(self.fields_)} }}'

    def build_data(self, data: Dict[str, Any]) -> str:
        """Returns a custom json dumped string of a dictionary.

        differences:
        - Without quotes surrounding each key value
        - Ignore null values

        e.g. {'title': 'hi', 'published': False, 'desc': None} -> {title: "hi", published: false}
        """
        strings = []
        dumps = json.dumps
        for key, value in data.items():
            if value is not None:
                strings.append(f'{key}: {dumps(value)}')

        return f'{{ {", ".join(strings)} }}'

    async def execute(self) -> Any:
        payload = self.build()
        return await self.engine.request('POST', '/', data=payload)
