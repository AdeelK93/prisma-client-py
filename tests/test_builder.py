# fmt: off
# I prefer this way of formatting
from typing import Dict, Any

import pytest
from prisma.builder import QueryBuilder
from prisma.errors import UnknownRelationalFieldError

from .utils import assert_query_equals


# TODO: more tests
# TODO: these tests should be schema agnostic
#       one schema should be used for all these tests
#       and only changed when needed for these tests
#       otherwise, simply adding a field to a model
#       will break these tests


def build_query(
    method: str,
    operation: str,
    arguments: Dict[str, Any],
    **kwargs: Any,
) -> str:
    return QueryBuilder(
        method=method,
        operation=operation,
        arguments=arguments,
        **kwargs,
    ).build_query()


def test_basic_building() -> None:
    assert_query_equals(QueryBuilder(
        operation='query',
        method='findUnique',
        model='User',
        arguments={'where': {'id': '1'}}
    ), '''
    query {
      result: findUniqueUser
      (
        where: {
          id: "1"
        }
      )
      {
        id
        name
      }
    }
    ''')
    assert_query_equals(QueryBuilder(
        operation='mutation',
        method='queryRaw',
        arguments={'where': {'id': '1'}}
    ), '''
    mutation {
      result: queryRaw
      (
        where: {
          id: "1"
        }
      )
    }
    ''')


def test_invalid_include() -> None:
    with pytest.raises(UnknownRelationalFieldError) as exception:
        QueryBuilder(
            operation='query',
            method='findUnique',
            model='User',
            arguments={
                'include': {
                    'hello': True,
                }
            }
        ).build_query()

    assert exception.match(
        'Field: "hello" either does not exist or is not a relational field on the User model'
    )


def test_include_no_model() -> None:
    with pytest.raises(ValueError) as exc:
        build_query(
            operation='mutation',
            method='queryRaw',
            arguments={'include': {'posts': True}}
        )

    assert exc.match('Cannot include fields when model is None.')


def test_include_with_arguments() -> None:
    assert_query_equals(QueryBuilder(
        operation='query',
        method='findUnique',
        model='User',
        arguments={
            'where': {'id': 1},
            'include': {'posts': {'where': {'id': 1}}}
        }
    ), '''
    query {
      result: findUniqueUser
      (
        where: {
          id: 1
        }
      )
      {
        id
        name
        posts(
          where: {
            id: 1
          }
        )
        {
          id
          created_at
          updated_at
          title
          published
          views
          desc
          author_id
        }
      }
    }
    ''')


def test_raw_queries() -> None:
    assert_query_equals(QueryBuilder(
        operation='mutation',
        method='queryRaw',
        arguments={
            'query': 'SELECT * FROM User where id = $1',
            'parameters': ["1263526"],
        }
    ), '''
    mutation {
      result: queryRaw
      (
        query: "SELECT * FROM User where id = $1"
        parameters: "[\\"1263526\\"]"
      )
    }
    ''')
