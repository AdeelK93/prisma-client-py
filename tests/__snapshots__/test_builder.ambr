# name: test_datetime_serialization_tz_aware
  '
  query {
    result: findUniquePost
    (
      where: {
        created_at: "1985-10-25T01:02:01+00:00"
      }
    )
    {
      id
      created_at
      updated_at
      title
      published
      views
      desc
      author_id
    }
  }
  '
---
# name: test_datetime_serialization_tz_unaware
  '
  query {
    result: findUniquePost
    (
      where: {
        created_at: "1985-10-26T00:01:01+00:00"
      }
    )
    {
      id
      created_at
      updated_at
      title
      published
      views
      desc
      author_id
    }
  }
  '
---
